### PR TITLE
Added ability to highlight searched word in hymn lyrics

### PR DESCRIPTION
**Pull request by jaypark**
Currently, the searched word gets highlighted in red for all occurrences of the word in the hymn detail page.  However, we're currently bypassing the hymn.formatText() call, which means that the hymn detail page is only showing the hymn lyrics only without any bolding of the refrain or without the credits.  We need to add the formatting back in. However, the challenge is that the highlighting code is looking through the hymn lyrics as a text/String. Once we use Text for the refrain part, it makes it difficult to parse through both String part and the Text part to highlight. Tried extracting the String part of the Hymn content (type Text), but seems rather challenging.  Could use some help here. Thanks.

### CODE REVIEW

File: A2N Hymnal/ContentView.swift
Comment (CCedricYoung) [inline A2N Hymnal/ContentView.swift:1]: @jaypark  what's with the enormous tab indents?

Comment (jaypark) [reply A2N Hymnal/ContentView.swift:1]: haha, ya, not sure what happened there... :) Looks OK on my local file...

Comment (CCedricYoung) [reply A2N Hymnal/ContentView.swift:1]: Okay I fixed it for you. You'll need to pull down the changes before making more.

Please set your indentation to `spaces` with `4` spaces indentation to make it easier for future, thanks!
image

Comment (jaypark) [reply A2N Hymnal/ContentView.swift:1]: yeah, my setting is set that way... weird.

**Comment by jaypark** [comment]
@paulchen95 @CCedricYoung : The PR/feature is now ready. Please review and approve. Thanks.

**Comment by jaypark** [comment]
Just verified that the highlight also works for Chinese! :) 
image

**Review by CCedricYoung — CHANGES_REQUESTED** [review]
@jaypark I think there is something wrong with your branch since the PR is showing all these new mp3 files and a couple of data changes.

**Comment by jaypark** [comment]
> @jaypark I think there is something wrong with your branch since the PR is showing all these new mp3 files and a couple of data changes.

It's because I rebased from main after some time of hiatus.  If you see this specific [commit](https://github.com/paulchen95/gphymnal/pull/47/commits/f6450096e69b33b4ee8c86b756069da0d5e6373c), you won't see those changes.

**Comment by jaypark** [comment]
Updated README as well.

File: A2N Hymnal/ContentView.swift [hunk 1]
```diff
@@ -39,12 +39,14 @@ struct ContentView: View {
         } //: NAVIGATION
         .sheet(isPresented: $showSettings, content: {
             SettingsView()
-            .environmentObject(viewModel)
-            .interactiveDismissDisabled(false)
+                .environmentObject(viewModel)
+                .interactiveDismissDisabled(false)
         })
+        //    DetailsView(hymn: Hymn)
```
Comment (paulchen95) [inline A2N Hymnal/ContentView.swift:45]: hi @jaypark wondering is there a reason the commented out "DetailsView(hymn: Hymn)" is added here?

File: A2N Hymnal/Hymn.swift [hunk 2]
```diff
@@ -60,25 +60,73 @@ struct Hymn: Identifiable {
...
+            // Get the punctuation characters at the end of the word, if any.
+            let suffixPunctuation = word.reversed().prefix { punctuationCharacterSet.contains($0.unicodeScalars.first!) }.reversed()
+            
+            // Highlight the word (or partial word) if exists.
+            if let range = trimmedWord.range(of: textToHighlight, options: .caseInsensitive) {
+                let beforeHighlight = trimmedWord[..<range.lowerBound]
+                let highlight = trimmedWord[range]
+                let afterHighlight = trimmedWord[range.upperBound...]
+                
+                // Re-stitch punctuation (if any)
```
Comment (paulchen95) [inline A2N Hymnal/Hymn.swift:112]: wow!

**Review by paulchen95 — COMMENTED** [review]
solid work @jaypark 
Just wondering if there is a way to say, I don't want those words to be highlighted anymore?
One common use case is that apparently people have been using A2N Hymnal to project directly to screen so people can sing to it (like a poor man's propre ... propre with songs that comes with it, what a concept 😄 )
And if I use search to find the hymn, the words will be highlighted red, but it'd be great to have an option now I know those words are the searched item, can I "dismiss" the highlight?  Then it'd not be awkward to show them on screen...

**Comment by paulchen95** [comment]
> Just verified that the highlight also works for Chinese! :) image

LOVE it!

File: A2N Hymnal/ContentView.swift (hunk 1 above)
Comment (jaypark) [reply A2N Hymnal/ContentView.swift:45]: o oops, i’ll remove that.

**Comment by jaypark** [comment]
> solid work @jaypark Just wondering if there is a way to say, I don't want those words to be highlighted anymore? One common use case is that apparently people have been using A2N Hymnal to project directly to screen so people can sing to it (like a poor man's propre ... propre with songs that comes with it, what a concept 😄 ) And if I use search to find the hymn, the words will be highlighted red, but it'd be great to have an option now I know those words are the searched item, can I "dismiss" the highlight? Then it'd not be awkward to show them on screen...

@paulchen95 : ya, that’d be nice. For now, though, they can easily go back, delete the search word and re-enter that song to address that as a workaround.  The search word is just a way to see what songs have that word.  Do you want to then wait until we come up with the Dismiss feature? Or, can that be the next feature add?  Not sure when I’ll have time to add that.  I don’t think your use case is all that common.  People generally know what they want to sing and select it.

**Comment by paulchen95** [comment]
> > solid work @jaypark Just wondering if there is a way to say, I don't want those words to be highlighted anymore? One common use case is that apparently people have been using A2N Hymnal to project directly to screen so people can sing to it (like a poor man's propre ... propre with songs that comes with it, what a concept 😄 ) And if I use search to find the hymn, the words will be highlighted red, but it'd be great to have an option now I know those words are the searched item, can I "dismiss" the highlight? Then it'd not be awkward to show them on screen...
> 
> @paulchen95 : ya, that’d be nice. For now, though, they can easily go back, delete the search word and re-enter that song to address that as a workaround. The search word is just a way to see what songs have that word. Do you want to then wait until we come up with the Dismiss feature? Or, can that be the next feature add? Not sure when I’ll have time to add that. I don’t think your use case is all that common. People generally know what they want to sing and select it.

alright @jaypark why don't you merge it to main?  Don't want to leave it hanging.  Also want to do some QA before releasing it out to the world.  May not have much time before ATTR #2 ...

**Review by paulchen95 — APPROVED** [review]
LGTM

File: A2N Hymnal/ContentView.swift (hunk 1 above)
Comment (jaypark) [reply A2N Hymnal/ContentView.swift:45]: in the next PR.

**Pull request merged**